README.md
# Draft.js exporter 🐍 - Markdown edition

[![PyPI](https://img.shields.io/pypi/v/draftjs_exporter_markdown.svg)](https://pypi.org/project/draftjs_exporter_markdown/) [![PyPI downloads](https://img.shields.io/pypi/dm/draftjs_exporter_markdown.svg)](https://pypi.org/project/draftjs_exporter_markdown/) [![Build status](https://github.com/thibaudcolas/draftjs_exporter_markdown/workflows/CI/badge.svg)](https://github.com/thibaudcolas/draftjs_exporter_markdown/actions) [![Coveralls](https://coveralls.io/repos/github/thibaudcolas/draftjs_exporter_markdown/badge.svg)](https://coveralls.io/github/thibaudcolas/draftjs_exporter_markdown) [![Total alerts](https://img.shields.io/lgtm/alerts/g/thibaudcolas/draftjs_exporter_markdown.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/thibaudcolas/draftjs_exporter_markdown/alerts/) [![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/thibaudcolas/draftjs_exporter_markdown.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/thibaudcolas/draftjs_exporter_markdown/context:python)

> Library to convert rich text from Draft.js raw ContentState to Markdown, based on [Draft.js exporter](https://github.com/springload/draftjs_exporter).
>
> 🚧 This is an experimental exporter with limited Markdown support – please use with caution.

## Usage

This package is a Markdown export configuration for the [Draft.js exporter](https://github.com/springload/draftjs_exporter). Specifically, it provides:

- A Markdown-friendly exporter engine, with fallbacks to HTML tags.
- Configuration for basic Markdown formatting.

First, install the package:

```sh
pip install draftjs_exporter_markdown
```

Then, to convert Draft.js content to Markdown:

And in Python:

```py
from draftjs_exporter.html import HTML
from draftjs_exporter_markdown import BLOCK_MAP, ENGINE, ENTITY_DECORATORS, STYLE_MAP

# Initialise the exporter.
exporter = HTML({
    # Those configurations are overridable like for draftjs_exporter.
    'block_map': BLOCK_MAP,
    'style_map': STYLE_MAP,
    'entity_decorators': ENTITY_DECORATORS,
    'engine': ENGINE,
})

markdown = exporter.render({
    'entityMap': {},
    'blocks': [{
        'key': '6mgfh',
        'text': 'Hello, world!',
        'type': 'unstyled',
        'depth': 0,
        'inlineStyleRanges': [],
        'entityRanges': []
    }]
})

print(markdown)
```

You can also run an example by downloading this repository and then using `python example.py`.

### Configuration

Please refer to the [Draft.js exporter configuration documentation](https://github.com/springload/draftjs_exporter#configuration).

### Supported Markdown formatting

The built-in configuration provides support for:

- Inline styles: bold, italic, strikethrough, code
- Blocks: paragraphs, heading levels, bullet and number lists, code blocks, blockquote
- Images, links, and horizontal rules

Contrary to the main Draft.js exporter,

- Nested / overlapping styles aren't supported.
- None of the content is escaped (HTML escaping is unnecessary for Markdown, and there is no Markdown escaping).

## Development

> Requirements: `virtualenv`, `pyenv`, `twine`

```sh
git clone git@github.com:thibaudcolas/draftjs_exporter_markdown.git
cd draftjs_exporter_markdown/

# Install dependencies
nvm install
npm install
# For tests and development in watch mode.
npm install -g nodemon

# Install the Python environment.
virtualenv .venv
source ./.venv/bin/activate
make init

# Install required Python versions
pyenv install --skip-existing 3.6.3
# Make required Python versions available globally.
pyenv global system 3.6.3

# Run the built-in example.
make dev
```

### Releases

Use `make release`, which uses [standard-version](https://github.com/conventional-changelog/standard-version) to generate the CHANGELOG and decide on the version bump based on the commits since the last release.

## Credits

View the full list of [contributors](https://github.com/thibaudcolas/draftjs_exporter_markdown/graphs/contributors). [MIT](LICENSE) licensed.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.4](https://github.com/thibaudcolas/draftjs_exporter_markdown/compare/v0.2.3...v0.2.4) (2022-04-05)

### Features

- **deps:** remove development dependencies from package extras ([5fcee41](https://github.com/thibaudcolas/draftjs_exporter_markdown/commit/5fcee415794d6288507b1902a4f4ba1100f88c72))
- **deps:** support draftjs_exporter v5.0.0 ([abb17db](https://github.com/thibaudcolas/draftjs_exporter_markdown/commit/abb17db33a2105f1b3e0a262d9d920193fc6ab46))
- **package:** clean up files packaged to PyPI ([a01c74f](https://github.com/thibaudcolas/draftjs_exporter_markdown/commit/a01c74f06664c30371443289eef9f47a2826e5cb))

### [0.2.3](https://github.com/thibaudcolas/draftjs_exporter_markdown/compare/v0.2.2...v0.2.3) (2020-06-27)

### Features

- **release:** publish package as a wheel ([fb898e8](https://github.com/thibaudcolas/draftjs_exporter_markdown/commit/fb898e8c524c130a747cb8968ce367650d2b988b))

### [0.2.2](https://github.com/thibaudcolas/draftjs_exporter_markdown/compare/v0.2.1...v0.2.2) (2020-05-17)

### Features

- **api:** remove Python 2.7 support ([6bd3988](https://github.com/thibaudcolas/draftjs_exporter_markdown/commit/6bd39882a00f6829df1fd260f41617559941a9d4))
- **deps:** add support for draftjs_exporter v4.0.0 ([b1a690a](https://github.com/thibaudcolas/draftjs_exporter_markdown/commit/b1a690a98c28028ef3675578c42a94e0a424bd48))

<a name="0.2.1"></a>

## [0.2.1](https://github.com/thibaudcolas/draftjs_exporter_markdown/compare/v0.2.0...v0.2.1) (2018-09-16)

### Bug Fixes

- **release:** set encoding when reading the README in setup.py ([#1](https://github.com/thibaudcolas/draftjs_exporter_markdown/issues/1)) ([e898ff3](https://github.com/thibaudcolas/draftjs_exporter_markdown/commit/e898ff3))

<a name="0.2.0"></a>

# [0.2.0](https://github.com/thibaudcolas/draftjs_exporter_markdown/compare/v0.1.0...v0.2.0) (2018-03-07)

### Features

- **exporter:** add support for real-world multi-block code blocks with wrapper ([849231e](https://github.com/thibaudcolas/draftjs_exporter_markdown/commit/849231e))
- **exporter:** remove support for code blocks language.

<a name="0.1.0"></a>

# 0.1.0 (2018-03-02)

First usable release! 🎉


md.py
# -*- coding: utf-8 -*-
from draftjs_exporter.html import HTML

from draftjs_exporter_markdown import BLOCK_MAP, ENTITY_DECORATORS, STYLE_MAP

if __name__ == '__main__':
    config = {
        'block_map': BLOCK_MAP,
        'style_map': STYLE_MAP,
        'entity_decorators': ENTITY_DECORATORS,
        'engine': 'draftjs_exporter_markdown.dom_markdown.DOMMarkwdown',
    }

    exporter = HTML(config)

    content_state = {
        "entityMap": {
            "0": {
                "type": "LINK",
                "mutability": "MUTABLE",
                "data": {
                    "url": "https://github.com/facebook/draft-js"
                }
            },
            "1": {
                "type": "LINK",
                "mutability": "MUTABLE",
                "data": {
                    "url": "https://facebook.github.io/react/docs/top-level-api.html#react.createelement"
                }
            },
            "2": {
                "type": "HORIZONTAL_RULE",
                "mutability": "IMMUTABLE",
                "data": {}
            },
            "3": {
                "type": "LINK",
                "mutability": "MUTABLE",
                "data": {
                    "url": "https://facebook.github.io/react/docs/jsx-in-depth.html"
                }
            },
            "4": {
                "type": "LINK",
                "mutability": "MUTABLE",
                "data": {
                    "url": "https://github.com/springload/draftjs_exporter/pull/17"
                }
            },
            "5": {
                "type": "IMAGE",
                "mutability": "IMMUTABLE",
                "data": {
                    "alt": "Test image alt text",
                    "src": "https://placekitten.com/g/300/200",
                    "width": 300,
                    "height": 200
                }
            },
            "6": {
                "type": "LINK",
                "mutability": "MUTABLE",
                "data": {
                    "url": "http://embed.ly/"
                }
            },
            "8": {
                "type": "EXAMPLE_MISSING",
                "mutability": "MUTABLE",
                "data": {
                    "url": "http://www.youtube.com/watch?v=feUYwoLhE_4",
                }
            },
        },
        "blocks": [{
            "key": "b0ei9",
            "text": "draftjs_exporter is an HTML exporter for Draft.js content",
            "type": "header-two",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [{
                "offset": 41,
                "length": 8,
                "key": 0
            }],
            "data": {}
        }, {
            "key": "74al",
            "text": "Try it out by running this file!",
            "type": "blockquote",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {
                "cite": "http://example.com/"
            }
        }, {
            "key": "7htbd",
            "text": "Features 📝🍸",
            "type": "header-three",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "32lnv",
            "text": "The exporter aims to provide sensible defaults from basic block types and inline styles to HTML, that can easily be customised when required. For more advanced scenarios, an API is provided (mimicking React's createElement) to create custom rendering components of arbitrary complexity.",
            "type": "unstyled",
            "depth": 0,
            "inlineStyleRanges": [{
                "offset": 209,
                "length": 13,
                "style": "CODE"
            }],
            "entityRanges": [{
                "offset": 209,
                "length": 13,
                "key": 1
            }],
            "data": {}
        }, {
            "key": "eqjvu",
            "text": " ",
            "type": "atomic",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [{
                "offset": 0,
                "length": 1,
                "key": 2
            }],
            "data": {}
        }, {
            "key": "9fr0j",
            "text": "Here are some features worth highlighting:",
            "type": "unstyled",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "2mhgt",
            "text": "Convert line breaks to <br>\nelements.",
            "type": "unordered-list-item",
            "depth": 0,
            "inlineStyleRanges": [{
                "offset": 23,
                "length": 4,
                "style": "CODE"
            }],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "f4gp0",
            "text": "Automatic conversion of entity data to HTML attributes (int & boolean to string, style object to style string).",
            "type": "unordered-list-item",
            "depth": 0,
            "inlineStyleRanges": [{
                "offset": 81,
                "length": 12,
                "style": "CODE"
            }, {
                "offset": 97,
                "length": 12,
                "style": "CODE"
            }],
            "entityRanges": [{
                "offset": 81,
                "length": 28,
                "key": 3
            }],
            "data": {}
        }, {
            "key": "3cnm0",
            "text": "Wrapped blocks (<li> elements go inside <ul> or <ol>).",
            "type": "unordered-list-item",
            "depth": 0,
            "inlineStyleRanges": [{
                "offset": 16,
                "length": 5,
                "style": "CODE"
            }, {
                "offset": 40,
                "length": 4,
                "style": "CODE"
            }, {
                "offset": 48,
                "length": 4,
                "style": "CODE"
            }],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "h5rn",
            "text": "With arbitrary nesting.",
            "type": "unordered-list-item",
            "depth": 1,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "5qfeb",
            "text": "Common text styles: Bold, Italic, Underline, Monospace, Strikethrough. cmd + b",
            "type": "unordered-list-item",
            "depth": 2,
            "inlineStyleRanges": [{
                "offset": 20,
                "length": 4,
                "style": "BOLD"
            }, {
                "offset": 26,
                "length": 6,
                "style": "ITALIC"
            }, {
                "offset": 34,
                "length": 9,
                "style": "UNDERLINE"
            }, {
                "offset": 45,
                "length": 9,
                "style": "CODE"
            }, {
                "offset": 56,
                "length": 14,
                "style": "STRIKETHROUGH"
            }, {
                "offset": 71,
                "length": 7,
                "style": "KEYBOARD"
            }],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "2ol8n",
            "text": "Overlapping text styles. Custom styles too!",
            "type": "unordered-list-item",
            "depth": 2,
            "inlineStyleRanges": [{
                "offset": 0,
                "length": 14,
                "style": "STRIKETHROUGH"
            }, {
                "offset": 12,
                "length": 4,
                "style": "BOLD"
            }, {
                "offset": 14,
                "length": 11,
                "style": "ITALIC"
            }],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "2lno0",
            "text": "#hashtag support via #CompositeDecorators.",
            "type": "unordered-list-item",
            "depth": 3,
            "inlineStyleRanges": [],
            "entityRanges": [{
                "offset": 21,
                "length": 20,
                "key": 4
            }],
            "data": {}
        }, {
            "key": "37n0m",
            "text": "Linkify URLs too! http://example.com/",
            "type": "unordered-list-item",
            "depth": 4,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "37n01",
            "text": "Depth can go back and forth, it works fiiine (1)",
            "type": "unordered-list-item",
            "depth": 2,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "37n02",
            "text": "Depth can go back and forth, it works fiiine (2)",
            "type": "unordered-list-item",
            "depth": 1,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "37n03",
            "text": "Depth can go back and forth, it works fiiine (3)",
            "type": "unordered-list-item",
            "depth": 2,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "37n04",
            "text": "Depth can go back and forth, it works fiiine (4)",
            "type": "unordered-list-item",
            "depth": 1,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "37n05",
            "text": "Depth can go back and forth, it works fiiine (5)",
            "type": "unordered-list-item",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "3tbpg",
            "text": " ",
            "type": "atomic",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [{
                "offset": 0,
                "length": 1,
                "key": 5
            }],
            "data": {}
        }, {
            "key": "5t6c9",
            "text": "For developers 🚀",
            "type": "header-three",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "2nb2a",
            "text": "Import the library",
            "type": "ordered-list-item",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "cfom5",
            "text": "Define your configuration",
            "type": "ordered-list-item",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "e2114",
            "text": "Go!",
            "type": "ordered-list-item",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "adt4j",
            "text": "Optionally, define your custom components.",
            "type": "ordered-list-item",
            "depth": 1,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "ed7hu",
            "text": "def blockquote(props):\n    block_data = props['block']['data']\n    return DOM.create_element('blockquote', {\n        'cite': block_data.get('cite')\n    }, props['children'])\n",
            "type": "code-block",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }, {
            "key": "1nols",
            "text": "Voilà!",
            "type": "unstyled",
            "depth": 0,
            "inlineStyleRanges": [],
            "entityRanges": [],
            "data": {}
        }]
    }
    markup = exporter.render(content_state)

    print(markup)
